src/lib/simulator/physics/Thermostat.h

```diff
@@ -0,0 +1,125 @@
+#pragma once
+
+#include "Particle.h"
+#include "container/container.h"
+#include "range/v3/algorithm.hpp"
+#include "range/v3/view/iota.hpp"
+#include "spdlog/spdlog.h"
+#include "utils/MaxwellBoltzmannDistribution.h"
+#include <bits/random.h>
+#include <cmath>
+#include <limits>
+
+template<const size_t DIMENSIONS>
+class Thermostat {
+ public:
+  /**
+     * @brief Constructor for the Thermostat class.
+     * @param t_init Initial temperature of the system.
+     * @param t_target Target temperature to maintain. Defaults to Tinit if not provided.
+     * @param delta_t Maximum temperature change allowed per application.
+     * @param nthermostat Number of steps after which the thermostat is applied.
+     * @param seed Seed for the random number generator.
+     */
+  Thermostat(double t_init, std::optional<double> t_target, std::optional<double> delta_t, unsigned int seed = std::random_device{}())
+      : t_init(t_init), seed(seed) {
+    this->t_target = t_target.has_value() ? *t_target : this->t_init;
+    this->delta_t = delta_t.has_value() ? *delta_t : std::numeric_limits<double>::infinity();
+
+    SPDLOG_DEBUG("Thermostat initialized with Tinit={}, Ttarget={}, deltaT={},  seed={}", t_init, *t_target, *delta_t, seed);
+  };
+
+  Thermostat() = default;
+
+  /**
+     * @brief Applies the thermostat to adjust particle velocities.
+     * @param particles Container of particles to apply the thermostat to.
+     */
+  void inline apply(container::ParticleContainer<DIMENSIONS> &particles) {
+    double current_temperature = calculateCurrentTemperature(particles);
+    SPDLOG_DEBUG("Current temperature calculated: {}", current_temperature);
+
+    double temp_difference = t_target - current_temperature;
+    if (std::abs(temp_difference) > delta_t) {
+      temp_difference = std::signbit(temp_difference) ? -delta_t : delta_t;
```

Comment:
you must also ensure that `delta_t` is a positive value, is that being done somewhere?  or you can simply use abs value irrespective of what user supplies